[APPSEC-59890] AAP Kubernetes Gateway Injector preview documentation (#33117)
DataDog/documentation: content/en/containers/kubernetes/appsec.md

---

> title: Kubernetes Application Security - Appsec Injector

Reviewer: ```suggestion
title: Kubernetes Application Security using Appsec Injector
```

Reviewer: Appsec is almost always spelled as AppSec.
Is there a reason we are spelling it as Appsec?

description: Automatically enable Application Security monitoring for ingress proxies and gateways in Kubernetes
aliases:
    - /agent/kubernetes/appsec
    - /security/application_security/setup/kubernetes/appsec-injector
further_reading:
- link: "/containers/kubernetes/apm/"
  tag: "Documentation"
  text: "Collect your application traces"
- link: "/containers/kubernetes/log/"
  tag: "Documentation"
  text: "Collect your application logs"
- link: "/security/application_security/setup/kubernetes/envoy-gateway"
  tag: "Documentation"
  text: "App and API Protection for Envoy Gateway"
- link: "/security/application_security/setup/kubernetes/istio"
  tag: "Documentation"
  text: "App and API Protection for Istio"
- link: "/security/default_rules/?category=cat-application-security"
  tag: "Documentation"
  text: "OOTB App and API Protection Rules"
- link: "/security/application_security/troubleshooting"
  tag: "Documentation"
  text: "Troubleshooting App and API Protection"
---

{{< callout url="#" btn_hidden="true" header="Appsec Injector is in Preview" >}}

> The Appsec Injector feature is in Preview. Use the following instructions to try the preview.

Reviewer: we should consider adding content to our Product Preview Program for Appsec Injector
https://www.datadoghq.com/product-preview/?product=security

Author: Yes it's in the making https://github.com/DataDog/corp-hugo/pull/22379/s :+1:

## Overview

The Datadog Appsec Injector automatically configures ingress proxies and gateways in your Kubernetes cluster to enable Application Security monitoring. This eliminates the need for manual proxy configuration and provides API-wide security coverage without modifying individual services or deploying tracers across your application fleet.

### What is the Appsec Injector?

The Appsec Injector is a Kubernetes controller that:
- **Automatically detects** supported proxies in your cluster (Envoy Gateway, Istio)
- **Configures proxies** to route traffic through an external Application Security processor
- **Enables threat detection** for all traffic passing through your ingress layer
- **Simplifies operations** through centralized configuration with Helm

### Supported proxies

- **Envoy Gateway**: Automatically creates `EnvoyExtensionPolicy` resources
- **Istio**: Automatically creates `EnvoyFilter` resources in the Istio system namespace

More proxies are available via manual installation on the global [setup page][10].

> 

Reviewer: ```suggestion
## Limitations

- Requires Datadog Cluster Agent 7.73.0 or later
- Supported proxy types: Envoy Gateway, Istio
- External processor must be manually deployed and scaled
- Cross-namespace service references require appropriate NetworkPolicy and ReferenceGrant configuration
- For specific proxy version compatibility, see:
  - [Envoy Gateway compatibility][8]
  - [Istio compatibility][9]
```

## Prerequisites

Before enabling the Appsec Injector, ensure you have:

1. A running Kubernetes cluster (version 1.20 or later)
2. [Datadog Cluster Agent 7.73.0+][1] installed and configured in your cluster
3. One or more supported proxies installed:
   - [Envoy Gateway][2]
   - [Istio][3]
4. [Remote Configuration][4] enabled to allow blocking attackers through the Datadog UI

## How it works

The Appsec Injector operates in **External Mode**, where a single Application Security processor deployment serves all gateway traffic in your cluster.

### Architecture

1. **External Processor Deployment**: You deploy a centralized Application Security processor as a Kubernetes Deployment with an associated Service.
2. **Automatic Proxy Detection**: The Appsec Injector controller watches for supported proxy resources in your cluster using Kubernetes informers.
3. **Automatic Configuration**: When proxies are detected, the injector creates the necessary configuration:
   - For Envoy Gateway: Creates `EnvoyExtensionPolicy` resources that reference the external processor service
   - For Istio: Creates `EnvoyFilter` resources in the Istio system namespace

> 4. **Traffic Processing**: Gateways route traffic to the external processor via the Kubernetes service for security analysis.

Reviewer: ```suggestion
- **External Processor Deployment**: You deploy a centralized Application Security processor as a Kubernetes Deployment with an associated Service.
- **Automatic Proxy Detection**: The Appsec Injector controller watches for supported proxy resources in your cluster using Kubernetes informers.
- **Automatic Configuration**: When proxies are detected, the injector creates the necessary configuration:
   - For Envoy Gateway: Creates `EnvoyExtensionPolicy` resources that reference the external processor service
   - For Istio: Creates `EnvoyFilter` resources in the Istio system namespace
- **Traffic Processing**: Gateways route traffic to the external processor via the Kubernetes service for security analysis.
```

Enable the injector and configure it to use your external processor service.

Configure the Appsec Injector using Helm values. Add the following to your `values.yaml`:

```yaml
datadog:
  appsec:
    injector:
      # Enable the Appsec Injector
      enabled: true

      # Enable automatic proxy detection (recommended)
      autoDetect: true

      # External processor configuration (required)
      processor:
        service:
          name: datadog-aap-extproc-service  # Required: name of the processor service
          namespace: datadog                  # Optional: defaults to Cluster Agent namespace
        port: 443
```

Install or upgrade the Datadog Helm chart:

```bash
helm upgrade -i datadog-agent datadog/datadog -f values.yaml
```

### Step 3: Enable auto-detection or specify proxies

**Auto-detection (recommended)**: Set `autoDetect: true` to automatically detect and configure all supported proxies in your cluster.

**Manual specification**: Alternatively, explicitly list the proxy types to configure:

```yaml
datadog:
  appsec:
    injector:
      enabled: true
      autoDetect: false
      proxies:
        - envoy-gateway
        - istio
```

> ### Step 4: verify installation

Reviewer: ```suggestion
### Step 4: Verify installation
```

```bash
kubectl logs -n datadog deployment/datadog-cluster-agent | grep appsec
```

Look for log messages indicating the injector has started and detected proxies.

#### Verify proxy configuration

For **Envoy Gateway**, check that `EnvoyExtensionPolicy` resources were created:

```bash
kubectl get envoyextensionpolicy -A
```

For **Istio**, check that `EnvoyFilter` resources were created:

```bash
kubectl get envoyfilter -n istio-system
```

#### Test traffic processing

Send requests through your gateway and verify they appear in the Datadog App and API Protection UI:

1. Navigate to [Security > Application Security][5] in Datadog
2. Look for security signals from your gateway traffic
3. Verify that threat detection is active

## Configuration reference

### Injector options

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `enabled` | Boolean | `false` | Enable or disable the Appsec Injector |
| `autoDetect` | Boolean | `true` | Automatically detect and configure supported proxies |
| `proxies` | Array | `[]` | Manual list of proxy types to configure. Valid values: `"envoy-gateway"`, `"istio"` |

> | `processor.service.name` | String | - | **Required.** Name of the external processor Kubernetes Service |

Reviewer: is there a reason a dash is here? You can also use Empty.

- Ensure [Remote Configuration][4] is enabled in your Datadog Agent

### External processor connection issues

**Symptom**: Gateways cannot reach the external processor.

**Solutions**:
- Verify the processor service name and namespace match your configuration
- Check for NetworkPolicy rules blocking cross-namespace traffic
- For Envoy Gateway: Ensure `ReferenceGrant` resources exist for cross-namespace service references
- Test DNS resolution from gateway pods: `nslookup datadog-aap-extproc-service.datadog.svc.cluster.local`
- Verify the processor port configuration matches the service definition

### RBAC permission errors

**Symptom**: Cluster Agent logs show permission denied errors.

**Solutions**:
- Verify the Cluster Agent ClusterRole includes permissions for:
  - `gateway.envoyproxy.io/envoyextensionpolicies` (create, update, delete, get, list, watch)
  - `networking.istio.io/envoyfilters` (create, update, delete, get, list, watch)
  - `gateway.networking.k8s.io/gateways` (get, list, watch)
  - `gateway.networking.k8s.io/gatewayclasses` (get, list, watch)
- Check that the ClusterRoleBinding references the correct service account

> ## Limitations

Reviewer: I would consider putting this section at the top of the doc so that users that do not meet the req's don't learn about that after trying to set this up. I'll add it to the top. If you agree, you can review it from here.
